src/mlpack/methods/decision_tree/best_binary_categorical_split_impl.hpp

```diff
@@ -0,0 +1,414 @@
+/**
+ * @file methods/decision_tree/all_categorical_split_impl.hpp
+ * @author Nikolay Apanasov (nikolay@apanasov.org)
+ *
+ * Implementation of the BestBinaryCategoricalSplit categorical split class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_DECISION_TREE_BEST_BINARY_CATEGORICAL_SPLIT_IMPL_HPP
+#define MLPACK_METHODS_DECISION_TREE_BEST_BINARY_CATEGORICAL_SPLIT_IMPL_HPP
+namespace mlpack {
+
+// Overload used in classification.
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename LabelsType,
+         typename WeightVecType>
+double BestBinaryCategoricalSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const size_t numCategories,
+    const LabelsType& labels,
+    const size_t numClasses,
+    const WeightVecType& weights,
+    const size_t minLeafSize,
+    const double minGainSplit,
+    vec& splitInfo,
+    AuxiliarySplitInfo& aux) 
+{
+  const size_t n = data.n_elem;
+  double bestFoundGain = std::min(bestGain + minGainSplit, 0.0);
+  bool improved = false;
+
+  // We are too small to split.
+  if (n < (minLeafSize * 2))
+    return DBL_MAX;
+
+  // Binary classification
+  if (numClasses == 2) {
+      // Order the categories of variable vₖ by their proportion in class C₁
+      // and map each categorical vₖ to its categorical rank
+      umat categoryCounts(numCategories, 2, fill::zeros);
+      vec categoryP(numCategories);
+      size_t totalCount;
+
+      for (size_t i = 0; i < n; ++i)
+          ++categoryCounts(data[i], labels[i]);
```

Comment:
No, the data values are important here, because those are the categories. We are computing `categoryP[i]` such that it holds the estimated probability of a training sample being in class one given that it is in the current tree node and its category is C_i. We are doing this because Theorem 4.5 of CART (and Proposition 7.1 of Ripley's _Pattern Recognition and Neural Networks)_ specifically order the samples by this value. To compute `categoryP[i]`, we need both the total count of the number of samples with category C_i, and the count of how many of these samples are in class one.